File: Binary Tree/python/binaryTree.py
from node import Node
import sys
sys.path.insert(0,'../../Linked Stack/python')
from stack import Stack 

from queue import Queue
from typing import Any

from depthFirstSearch import DepthFirstSearch
class BinaryTree:
    def __init__(self) -> None:
        self.root:Node = None
        
    def insert(self, value:Any)->None:
        newNode = Node(value=value)
        if(self.root == None):
            self.root = newNode
            return
        q = Queue()
        temp:Node = self.root
        q.put(item=temp)
        
        while(not q.empty()):
            temp = q.get()
            if(temp.left == None):
                temp.left = newNode
                return
            else:
                q.put(temp.left)
            
            if(temp.right == None):
                temp.right = newNode
                return
            else:
                q.put(temp.right)
    
    def inOrder(self)-> None:
        DepthFirstSearch(self.root).displayInOrder()
                
    def preOrder(self) ->None:
        DepthFirstSearch(self.root).displayPreOrder()
        
    def postOrder(self) ->None:
        DepthFirstSearch(self.root).displayPostOrder()
    
    def remove(self, value:Any) ->None:
        if(self.root == None): return
        
        if(self.root.left == None and self.root.right == None):
            if(self.root.value == value):
                self.root = None
                
            return
        
        q:Queue = Queue()
        temp:Node = self.root
        q.put(temp)
        keyNode:Node = None
        
        while(not q.empty()):
            temp = q.get()
            
            if(temp.value == value):
                keyNode = temp
            
            if(temp.left != None):
                q.put(temp.left)
            if(temp.right != None):
                q.put(temp.right)
            
        if(keyNode != None):
            keyNode.value = temp.value
            self.__remove(temp)
            
    def __remove(self,delNode:Node) -> None:
       
        
        s = Stack()
        temp:Node = self.root
        s.push(item=temp)
        
        while(not s.isEmpty()):
            parint = temp;
            temp = s.pop()
            
            if(temp == delNode):
                if(parint.left != None and parint.left == delNode):
                    parint.left = None
                    return
                else:
                    parint.right =  None
                    return
            
            
            if(temp.left != None):
                s.push(temp.left)
            if(temp.right != None):
                s.push(temp.right)
            
            
    
        

File: Binary Tree/dart/binary_tree.dart
import 'node.dart';
import "../../Queue/dart/queue.dart";
import 'depth_first_search.dart';
import '../../Stack/dart/stack.dart';

class BinaryTree<T> {
  Node<T>? root;
  BinaryTree() {
    root = null;
  }

  void insert(T value) {
    Node<T> newNode = Node(value);
    if (root == null) {
      root = newNode;
      return;
    }

    Queue<Node<T>> queue = Queue<Node<T>>();

    Node<T> temp = root!;

    queue.enqueue(temp);

    while (!queue.isEmpty()) {
      temp = queue.getFront();
      queue.dequeue();

      if (temp.left == null) {
        temp.left = newNode;
        return;
      } else {
        queue.enqueue(temp.left!);
      }

      if (temp.right == null) {
        temp.right = newNode;
        return;
      } else {
        queue.enqueue(temp.right!);
      }
    }
  }

  void inOrder() {
    DepthFirstSearch<T>(root!).displayInOrder();
  }

  void preOrder() {
    DepthFirstSearch<T>(root!).displayPreOrder();
  }

  void postOrder() {
    DepthFirstSearch<T>(root!).displayPostOrder();
  }

  void remove(T value) {
    if (root == null) return;

    if (root!.left == null && root!.right == null) {
      if (root!.value == value) {
        root = null;
        return;
      } else
        return;
    }

    Queue<Node<T>> queue = Queue<Node<T>>();

    Node<T> temp = root!;
    Node<T>? keyNode = null;

    queue.enqueue(temp);

    while (!queue.isEmpty()) {
      temp = queue.getFront();
      queue.dequeue();

      if (temp.value == value) {
        keyNode = temp;
      }
      if (temp.left != null) {
        queue.enqueue(temp.left!);
      }
      if (temp.right != null) {
        queue.enqueue(temp.right!);
      }
    }

    if (keyNode != null) {
      keyNode.value = temp.value;
      _deleteNode(temp);
    }
  }

  void _deleteNode(Node<T> delNode) {
    Stack stack = Stack();

    Node<T> temp = root!;
    Node<T>? parint = null;
    stack.push(temp);

    while (!stack.isEmpty()) {
      parint = temp;
      temp = stack.pop();

      if (temp == delNode) {
        if (parint.left != null && parint.left == delNode) {
          parint.left = null;

          return;
        }
        parint.right = null;
        return;
      }

      if (temp.left != null) {
        stack.push(temp.left);
      }
      if (temp.right != null) {
        stack.push(temp.right);
      }
    }
  }
}
